073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,364 @@
+
+### New trust-state Kubernetes Secret
+
+A new Kubernetes Secret is introduced which is updated by the cluster operator in its role of managing the cluster (specifically which issued certificates are stored where and rolling pods to use/trust those certificates) to reflect whether the individual CA certificates are trusted and/or in use in the cluster.
+
+```yaml
+kind: Secret
+metdata:
+  name: <CLUSTER_NAME>-cluster-ca-trust-state # e.g. foo-cluster-ca-trust-state
+data:
+  ${fingerprint}: ${state}
+```
+
+The `fingerprint` is the fingerprint/thumbprint of the certificate. I.e. the SHA-1 hash of the ASN.1 DER-encoded form of the X509 certificate, which is a commonly used way of identifying certificates using common tooling.
+The `state` can be any of the states of the CA certificate trust state machine described below.
+
+### CA certificate state machine
+
+The "CA certificate trust state machine" has the following states:
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_UNUSED // trusted everywhere it needs to be, but no EE certificates yet issued with this CA certificate
+   |
+   v
+  TRUSTED_IN_USE // trusted everywhere, EE certificates issued using this CA certificate
+   |
+   v
+  PHASE_OUT // trusted everywhere, but new end-entity certificates will not be issued using this CA certificate
+   |
+   v
+  ZOMBIE // still trusted by some component, but with no issued end-entity certificates in current use.
+```
+
+### New trusted-certs Kubernetes Secret
+
+A new Kubernetes Secret is introduced which is updated by the cluster operator in its role as the receiver of issued certificates to include the CA certificates that the components should currently trust.
+This Kubernetes Secret will be mounted into Kubernetes Pods to use a TrustStore.
+
+```yaml
+kind: Secret
+metdata:
+  name: <CLUSTER_NAME>-cluster-ca-trusted-certs # e.g. foo-cluster-ca-trusted-certs
+type: strimzi.io/trust
+data:
+   ${fingerprint}.crt: <PEM encoded CA certificate>
+```
+
+The `fingerprint` is the fingerprint/thumbprint of the certificate. I.e. the SHA-1 hash of the ASN.1 DER-encoded form of the X509 certificate, which is a commonly used way of identifying certificates using common tooling.
+
+### Example flows
+
+When a new cluster is first created the flow is (new/altered steps in bold):
+
+1. Either user or CaReconciler creates the Cluster CA certificate and key and stores them in the <CLUSTER_NAME>-cluster-ca and <CLUSTER_NAME>-cluster-ca-cert Kubernetes Secrets.
+2. __The CaReconciler takes the Cluster CA certificate and places it in the trust-state Kubernetes Secret with the state UNTRUSTED.__
+3. The CaReconciler generates the Cluster Operator certificate and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret.
```

Comment:
If there are multiple CA certs in the trust-state how does it decide which one to use for issuing new EE certs like this one? In particular during cluster bootstrapping like this it's OK to use one that's `UNTRUSTED`, but once a cluster is running we use one that's `TRUSTED_UNUSED`. And I guess if we needed a new cert for a newly added broker we'd select one that was `TRUSTED`?

Does the CaReconciler 'know' which CA cert to use because of in-memory state (i.e. state that's been built in-memory by the prior steps in this reconciliation)? In which case we need to consider the edge cases that come about if the CO is restarted part way through and that in-memory state is lost. 

Or does it decide based purely on the persistent state (Secrets)? In which case it needs to be able to figure out the signing cert and key based purely on the Secrets. I think that needs to be explained in more detail. For example, does anything guarantee that there is only ever a single cert that's `TRUSTED`? If not, which will be used?

Reply:
The Strimzi CA uses the certificate and key stored in the <CLUSTER_NAME>-cluster-ca and <CLUSTER_NAME>-cluster-ca-cert Kubernetes Secrets when issuing new EE certs. So it doesn't rely on either the in-memory state or the persistent state. The persistent state is used to make sure that no new EE certs are issued until all components trust the most recently added CA cert (apart from as you say during the cluster bootstrapping). In practice I think there should only be a single cert `TRUSTED_IN_USE`, but I don't think we need that guarantee for the flow to work.